demo/angular-ngFor.ts

```diff
@@ -1,66 +1,122 @@
+  @ViewChildren("gridStackItem") gridstackItems: QueryList<any>;
   @Input() public items: GridStackWidget[] = [
-     { x: 0, y: 0, w: 9, h: 6},
-     { x: 9, y: 0, w: 3, h: 3},
-     { x: 9, y: 3, w: 3, h: 3},
-   ];
-   private grid: GridStack;
- 
-   constructor() {}
- 
-   // wait until after DOM is ready to init gridstack - can't be ngOnInit() as angular ngFor needs to run first!
-   public ngAfterViewInit() {
-     this.grid = GridStack.init({
-       cellHeight: 70,
-     });
-   }
- 
-   /**
-    * this would be easier with addWidget(), removeWidget(), update() but simulating angular change detection instead...
-    */
-   public add() {
-     // this SHOULD trigger ngOnChanges() but not seeing it... and doing ngDoCheck() seem extreme ?
-     // https://www.reddit.com/r/angular/comments/azjefs/change_detection_for_arraysobjects/
-     // https://angular.io/guide/lifecycle-hooks#docheck
-     this.items = [...this.items, {x: 1, y: 6, w: 3}];
-     // this.items.push({x: 1, y: 6, w: 3});
-   }
-   public delete() {
-     this.items.pop(); // todo
-   }
-   public change() {
-     this.items[0].w = 1; // todo
-   }
- 
-   public ngOnChanges(changes: SimpleChanges) {
-     if (changes.items) {
-       // TODO: ... figure what is new and call makeWidget(), old -> removeWidget(el,false) and changed -> update()
-       console.log('items changed');
-     }
-   }
- }
- 
\ No newline at end of file
+    { x: 0, y: 0, w: 1, h: 1 },
+    { x: 1, y: 1, w: 1, h: 1 },
+    { x: 2, y: 2, w: 1, h: 1 },
+  ];
+
+  private grid: GridStack;
+  private widgetToMake: Subject<{
+    action: "add" | "remove" | "update";
+    id: number;
+  }> = new Subject(); // consider to use a statemanagement like ngrx component store to do this
+
+  constructor() {}
+
+  // wait until after DOM is ready to init gridstack - can't be ngOnInit() as angular ngFor needs to run first!
+  public ngAfterViewInit() {
+    this.grid = GridStack.init({
+      alwaysShowResizeHandle:
+        /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
+          navigator.userAgent
+        ),
+      margin: 5,
+      float: true,
+    });
+
+    // To sync dom manipulation done by Angular and widget manipulation done by gridstack we need to zip the observables
+    zip(this.gridstackItems.changes, this.widgetToMake).subscribe(
+      ([changedWidget, widgetToMake]) => {
+        if (widgetToMake.action === "add") {
+          this.grid.makeWidget(`#${widgetToMake.id}`);
+        } else if (widgetToMake.action === "remove") {
+          const removeEl = this.grid
+            .getGridItems()
+            .find((el) => el.id === `${widgetToMake.id}`);
+          this.grid.removeWidget(removeEl);
+        }
+      }
+    );
+  }
+
+  /**
+   * CRUD operations
+   */
+  public add() {
+    this.items = [...this.items, { x: 3, y: 0, w: 3 }];
+    this.widgetToMake.next({ action: "add", id: this.items.length - 1 });

```

Comment:
Yes. The problem was to sync dom changes and gridstack changes, which I did through the observables.

Reply:
re-reading this again... how is calling this.widgetToMake.next() any better than calling this.grid.makeWidget() directly ?
I don't see the benefit of all that convulated subject. reality is the person who makes changes to the array STILL has to (in)directly call makeWidget() as it doesn't happen automatically... what I was trying to avoid with ngOnChanges

Reply:
If you call makeWidget directly in 90% of the times the dom changes are not done yet. This will fail making the widget.
Because of this we need to wait for the dom changes to finish and the make the widget. To sync this the easiest way is to have an Observable and „zip“ it which means: wait for both observables to emit a value and then execute the code in subscribe.

If you like, give it a try with ngOnChanges, but I suspect this won‘t work due to timing issues and ngOnChanges is only triggered when passing in values from outside the component.

Reply:
but we could wait for the DOM changes to emit, then makeWidget (or remove info) in that event triggering, no real need to create a fake widgetToMake info that user has to know about... I'll give it a try when I find time. that info couldbe gotten from the first param.

Reply:
I believe this is the better way - no special command, track any external changes and updates internal list to match as well
https://github.com/gridstack/gridstack.js/blob/master/demo/angular/src/app/ngFor.ts

Reply:
From my perspective this is less robust and harder to understand. Using setTimeout to avoid ExpressionChangedAfterItHasBeenCheckedError is a hack. This should be solved properly. Furthermore you are mutating the existing array instead of generating new ones which circumvents the angular way of detecting changes.

Reply:
| Using setTimeout to avoid ExpressionChangedAfterItHasBeenCheckedError is a hack

Agree. this only happens for newly created items (not 'change'), and could be solved by GS not calling on 'added' until we return to the main event loop. I tried using this.zone.runOutsideAngular() but didn't help.

| you are mutating the existing array instead of generating new ones which circumvents the angular way of detecting changes.

So I had creating a new array (for add at least) but noticed it worked perfectly fine modifying the content instead, which must have something todo with trackBy using actual unique id and not the index (which doesn't stay the same across calls).

You version required the user to know to call special command to update GS (mine avoids that) and I believe re-created them widgets on each array change, rather than update what changed - this could be a big performance inpact once you have real content inside them... your trackBy index could be random depending on how you insert new items, or whihc you delete. mine isn't index based.

Reply:
keep in mind that `.on('change added')` is optional (works without it) and wasn't something you even did in your example (the items and GS nodes will get out of sync)

Reply:
Also your `(widgetToMake.action === "remove")` fails as the element has been removed by Angular by the time you look for it...

Reply:
👍🏻